include/deal.II/fe/fe_values_coupling.h

```diff
@@ -0,0 +1,557 @@
+     *
+     * @dealiiRequiresUpdateFlags{update_gradients}
+     */
+    template <typename Number>
+    void
+    get_function_gradients(
+      const ReadVector<Number> &                   fe_function,
+      std::vector<solution_gradient_type<Number>> &gradients) const;
+
+    /**
+     * Same as above, but using a vector of renumbered local degree-of-freedom
+     * gradients. In other words, instead of extracting the nodal values of the
+     * degrees of freedom located on the current cell from a global vector
+     * associated with a DoFHandler object (as the function above does), this
+     * function instead takes these local nodal values through its first
+     * argument.
+     *
+     * @param[in] dof_values A vector of local nodal values. This vector must
+     *   have a length equal to the size of the dof renumbering vector.
+     *
+     * @param[out] gradients A vector of gradients of the given finite element
+     *   field, at the renumbered quadrature points on the current object.
+     *
+     * @tparam InputVector The @p InputVector type must allow creation of an
+     *   ArrayView object from it; this is satisfied by the `std::vector` class,
+     *   among others.
+     */
+    template <class InputVector>
+    void
+    get_function_gradients_from_local_dof_values(
+      const InputVector &dof_values,
+      std::vector<solution_gradient_type<typename InputVector::value_type>>
+        &gradients) const;
+
+  private:
+    /**
+     * Helper function that constructs a unique name for a container, based on a
+     * string prefix, on its size, and on the type stored in the container.
+     */
+    template <typename Number>
+    std::string
+    get_unique_container_name(const std::string &prefix,
+                              const unsigned int size,
+                              const Number &     exemplar_number) const;
+
+    /**
+     * Produce an inner vector compatible with the inner view, after copying the
+     * values with the correct numbering from the outer vector.
+     */
+    template <typename InputVector>
+    const InputVector &
+    outer_to_inner_dofs(const InputVector &outer_vector) const;
+
+    /**
+     * Produce an inner vector compatible with the inner view, and zero out its
+     * entries if necessary.
+     */
+    template <typename ValueType>
+    std::vector<ValueType> &
+    outer_to_inner_values(std::vector<ValueType> &outer_values) const;
+
+    /**
+     * Return the outer argument renumbered according to the quadrature
+     * renumbering. The values in the inner values are copied to the right
+     * position in the outer vector.
+     */
+    template <typename ValueType>
+    void
+    inner_to_outer_values(const std::vector<ValueType> &inner_values,
+                          std::vector<ValueType> &      outer_values) const;
+
+    /**
+     * Store a reference to the underlying view.
+     */
+    const ViewType &view;
```

Comment:
Just make a copy. We don't generally create views as objects with storage, but only as temporaries.

Reply:
Actually, on second thought, we generally take them from a cache. So perhaps this design is right. Maybe say that `FEValues` keeps a cache of these objects and so we can safely keep a reference.

Reply:
copying them is also forbidden.